Created logger

ynaydenov-qb opened the pull request:
Finished the last requirement, now I am ready for feedback and to refactor/improve the code where needed

ynaydenov-qb committed:
Create basic Apollo server as starting point

ynaydenov-qb committed:
Created models and corresponding files for storage

ynaydenov-qb committed:
Added schema and resolvers

Added schema for the models and completed the according resolvers for all queries and mutations. Tested manually, didn't add comments.

ynaydenov-qb committed:
Added comments and did some more manual testing on normal behavior

ynaydenov-qb committed:
Added authentication which checks for a token in the request headers

ynaydenov-qb committed:
Created logger

ynaydenov-qb committed:
Added checks with error messages to the mutations

astanilov commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+node_modules/
+src/logs/
```
🔥

Not a problem right now, but for the sake of project security, it's a must to also include any real `.env` files in the ignore also. By real I mean those that will be used on prod/staging and dev. 

It's okay to have a `sample.env` of some sort, with fake data, so that new users know what to provide as variables.

Pushing sensitive data in the repo is considered bad practice and can lead to project compromise.

Totally not a blocker, just wanted to let you know, in case you are not aware 😇

astanilov replied:
also we should include `.dist` or `.build` folder if our build process produces any of those

ynaydenov-qb replied:
Thank you for the explanation, I will remember it for future projects.

astanilov commented on src/app.ts:
```diff
@@ -0,0 +1,35 @@
+
+dotenv.config();
+
+const authenticationContextMiddleware = ({ req }: { req: any }) => {
```
What do you think if we export the middleware in it's own folder and specific file like `middleware/authentication.ts`?
It will be easier to scale and to search the code.

ynaydenov-qb replied:
I think it is great idea and I will do it, it really will allow for greater scalability.

ynaydenov-qb replied:
Done.

astanilov commented on src/logger.ts:
```diff
@@ -0,0 +1,19 @@
+import { createLogger, format, transports } from "winston";
+import path from "path";
+
+const logPath = path.join(__dirname, "logs", "server.log");
```
Looks good! Have you explored the `logs rotation` technique? Do you think we might benefit from it here?

ynaydenov-qb replied:
I did a SPIKE and it turned out that the logs rotation technique has a lot of benefits such as having daily log files, automatic cleanup after a period of time and also managing the maximum size of a log file, therefore I decided to implement it also in this project and altered the corresponding code. Now it is working.

astanilov replied:
great job with that, thank you!

astanilov commented on src/models.ts:
```diff
@@ -0,0 +1,58 @@
+export class Book {
```
a good way to structure the code and have increased readability/maintainability is to have models as a folder `models` and split every model in it's own file

ynaydenov-qb replied:
Thanks for the advice, I updated the models such that each of them is in a separate file in a models folder.

astanilov commented on src/models.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+}
+
+export class LendingRecord {
```
I think LendingRecord should also have an id, not going to use it for now, but all entries in most of the DBs have unique keys(id), and if we want to show those records somewhere in the UI, might be good to have the ID for CRUD operations or such

ynaydenov-qb replied:
I added it, your point makes sense

astanilov commented on src/schema.ts:
```diff
@@ -0,0 +1,45 @@
+import { gql } from "apollo-server";
+
+export const typeDefs = gql`
```
can we try splitting the schema in to smaller entries in their own files and unite them here?

ynaydenov-qb replied:
Good remark, I did it, thanks

astanilov commented on src/schema.ts:
```diff
@@ -0,0 +1,45 @@
+  type LendingRecord {
+    customer: Customer!
+    lentDate: String!
+    returnDate: String
```
won't the return date be also optional?

ynaydenov-qb replied:
From my understanding of GraphQL if a field doesn't have the "!" it can be nullable which in my case serves as optional, so if we don't have a return date it is just null, which can be used to check if a book is returned later.

astanilov replied:
ah yeah that's right! I was looking at the .ts extension of the file and was thinking about TS type here. nvm my comment, but thanks for taking the time to explain it 🥳

astanilov commented on src/data/booksDataSource.ts:
```diff
@@ -0,0 +1,50 @@
+// Path to JSON file where books are stored
+const booksDataPath = path.join(__dirname, "booksData.json");
+
+export class BooksDataSource {
```
DataSources should also have their own folder, somewhere next to models and resolvers, while keeping `data` only for the JSON files with the actual data.

ynaydenov-qb replied:
I separated them, the JSON files remained in the data folder, while the data sources where moved to a separate folder called dataSources. All paths have been fixed to point to the right files.

astanilov commented on src/resolvers.ts:
```diff
@@ -0,0 +1,147 @@
+  },
+
+  Mutation: {
+    // Add a new book, the isLent field is set to false by default
```
good job with the comments 🥳

ynaydenov-qb replied:
Thanks!

astanilov commented on src/resolvers.ts:
```diff
@@ -0,0 +1,147 @@
+      _: any,
+      { id, title, author }: { id: string; title: string; author: string }
+    ): Book => {
+      const existingBook = booksDataSource.getBookById(id);
```
Hmm I think that ID should not be provided when creating a new Book/Customer/Records

If we think a bit globally and pretend that data comes from the UI, if we are creating a new Book, the ID will not be known on the UI and should not be generated there.

For speed and reliability and security the BE(or the DB actually) should have a mechanism of generating unique IDs for every entry in every table.

Do you think we can have that here?

ynaydenov-qb replied:
You are right, ID should not be provided by user. Normally the ID is autogenrated by the database using a strategy (increment, etc.) however in our case we don't have a database to provide us with this strategy. After a research I decided to use the 'uuid' package in order to generate UUIDs. These UUIds are designed to be globally unique, therefore no two instances of any model will have the same id.

astanilov replied:
awesome! that is exactly what we use here for similar cases 🔥

astanilov commented on src/schema.ts:
```diff
@@ -0,0 +1,45 @@
+    title: String!
+    author: String!
+    isLent: Boolean!
+    currentLendee: Customer
+    dueDate: String
+    lendingHistory: [LendingRecord!]!
```
I guess those can be marked as optional, as not every Book will be lent at the time of its creation. 

Also, do you think `dueDate` should be saved for the Book entry, or for LendingRecord entry? I think the Book itself does not have a dueDate(except if it is not food and has expiry date), but more like the dueDate should be filled when a Customer takes a Book and we create a LendingRecord that connects Book to Customer

ynaydenov-qb replied:
I agree with you, I will remove the dueDate here and leave it only in the lendingRecord.

astanilov commented:
Wow I didn't expect that you will provide everything so fast! 🥳

Something that is more our mistake, was that we should have stated in the task that it is better to iteratively push code to the repo. Splitting the task into smaller ones, and providing just the required code for the task. It is much easier to follow your own progress, it's easier to review small chunks of code also. And last, but not least, if some change needs to be reverted - it can also be done iteratively.

Don't take this comment as a critique, more like a guidance something that is part of our daily life.

Also, most of the repos you will work with will have a template for the PR info. So just FYI it's a good practice to have more descriptive information provided. For example:
1. Related ticket/task(we don't have such currently)
2. What has changed/what was added and why
3. Screenshots/Recordings also provide a lot of context
4. How the reviewer can test your changes

Sorry for not providing that info upfront

ynaydenov-qb committed:
Exported the authentication middleware in another folder

Exported the authentication middleware in another folder named middleware, this would allow for scalability in later stages if more middleware functions are needed.

ynaydenov-qb committed:
Added log rotation

Added log rotation in order to have daily logs, automatic cleanup and be able to manage the logs size

ynaydenov-qb committed:
Exported every model to its own file in a models folder

ynaydenov-qb committed:
Added autogeneration for the models' id using the "uuid" package

ynaydenov-qb committed:
Split the schema into smaller ones in the folder schemas

Split the schema into smaller ones in the folder schemas and finally combined them into the file schema.ts

ynaydenov-qb committed:
Separated the data sources and the data into two folders

ynaydenov-qb committed:
remove dueDate field from a book as it is redundant

ynaydenov-qb commented:
Thanks for all the info, I know that I have to split the task into smaller ones and I did so ( I have 6 branches on this repo), but I didn't want to spam you with a pull requests and just left the final one.

astanilov commented on src/dataSources/booksDataSource.ts:
```diff
@@ -0,0 +1,50 @@
+import { Book } from "../models/book";
+import * as fs from "fs";
```
you don't need to import `* as fs` which just imports everything and names it `fs`. You can simply do that with
```suggestion
import fs from "fs";
```

astanilov replied:
This applies for too all other places

ynaydenov-qb replied:
I did it, thanks for the advice!

asynchroza commented on package.json:
```diff
@@ -0,0 +1,28 @@
+  "license": "MIT",
+  "dependencies": {
+    "@graphql-tools/merge": "^9.0.6",
+    "@types/node": "^22.4.1",
```
Not a major issue here but we tend to lock versions in order to avoid potential updates from breaking our code.

ynaydenov-qb replied:
I locked the versions in the newest package.json file, thanks

asynchroza commented on src/app.ts:
```diff
@@ -0,0 +1,16 @@
+
+server.listen().then(({ url }) => {
+  logger.info(`Server started at ${url}`);
+  console.log(`Server started at ${url}`);
```
nit: You don't need the `console.log` here anymore.

ynaydenov-qb replied:
Absolutely right, I removed it.

asynchroza commented on src/logger.ts:
```diff
@@ -0,0 +1,30 @@
+
+const logDir = path.join(__dirname, "logs");
+
+export const logger = createLogger({
```
If I'm not mistaken, your logger doesn't pipe its output to stdout when running locally.
In order to avoid having to use both console.logs and custom loggers, what we do is check whether `process.env.NODE_ENV` is "development" and add a transport which pipes the log to stdout. 

Another thing is that we would usually prefer to store the logs in a JSON format in production environments. This allows for easier querying down the line.

ynaydenov-qb replied:
I added a `NODE_ENV` variable in the environment and also altered the logger to output logs to the console while in development.

mbozhilov-qb replied:
You shouldn't be adding the NODE_ENV environment variable by yourself. Usually, your environment will set this up for you.

mbozhilov-qb replied:
Usually build scripts will set it to "production" for you and dev environment tools such as hot reload runners such as `ts-node-dev` will set it as "development".

astanilov commented:
Excellent work on the latest changes 🔥

Things are looking way more concise and maintainable
![image](https://github.com/user-attachments/assets/972a6952-9a9d-4bb9-9a2e-ce1a80d712f9)


Two things come to my mind, and we should have probably included them as part of the task:
**1.** I saw 2-3 places where you import something and you don't use it in the module/file. Not a big problem, but we strive to have a meaningful code, without extra dependancies. 
Instead of just cleaning those manually, wha do you think if we implement ESlint and Prettier in the project? You will see how those things work and those two go hand by hand in 9 of 10 projects you will come across. 
We will let you choose what rules to apply, but you can start with those official [Airbnb](https://www.npmjs.com/package/eslint-config-airbnb-base) rules.
Also let's include an aslant task in the scripts part of package.json so we can enforce aslant rules on the whole project.
You can also dig how to set up aslant and prettier with VSCode and add the configuration file to the project as well

**2.** A tedious task, but it happens to every developer's life at some point - writing documentation. Let's make use of that README.md ad add a bit of info on how to set up the project, and how to run it and test it. Don't need to be do exhaustive.

asynchroza commented on src/middleware/authentication.ts:
```diff
@@ -0,0 +1,22 @@
+import { logger } from "../logger";
+import * as dotenv from "dotenv";
+
+dotenv.config();
```
I recommend moving this to the top of your entry point file where the server is initialized. Currently, all your services and middleware that rely on environment variables are dependent on the initialization of the authentication middleware.

ynaydenov-qb replied:
Well noted, I didn't think of that. I moved this to the top of app.ts

asynchroza commented on src/middleware/authentication.ts:
```diff
@@ -0,0 +1,22 @@
+
+dotenv.config();
+
+export const authenticationContextMiddleware = ({ req }: { req: any }) => {
```
We should strive to avoid using any. You can probably infer what arguments are being passed to `authenticationContextMiddleware` by hovering over this line in `  context: authenticationContextMiddleware,` in app.tsx

ynaydenov-qb replied:
That is my bad, I knew that the type was Request, but forgot to add it. Now it is added.

asynchroza commented on src/middleware/authentication.ts:
```diff
@@ -0,0 +1,22 @@
+  const expectedHeader: string = process.env.AUTH_TOKEN!;
+
+  if (authHeaderUser !== expectedHeader) {
+    logger.error(`Unauthorized access attempt: ${req.method} ${req.url}`);
```
Usually logs cost money and we tend to avoid logging every request. Instead, we generate arbitrary ids which helps us trace requests down the line.

ynaydenov-qb replied:
I didn't understand what you mean exactly, can you give me some more explanations.

mbozhilov-qb replied:
TLDR:

Usually, in APIs, you would pass the reference to the database, datasources, services and loggers as context. A request receives said context and can do something of the sorts of:

```typescript
ctx.logger.info("Hello World"); // which will log { "message": "Hello World", "requestId": "some-arbitrary-id" }
```

This id is used to track the flow of a request. 

Let's skip the implementation for the moment being but it's good to keep it in mind for future reference.

asynchroza commented on src/middleware/authentication.ts:
```diff
@@ -0,0 +1,22 @@
+    throw new AuthenticationError("Unauthorized");
+  }
+  logger.info(
+    `Received request: ${req.method} ${req.url} with headers: ${JSON.stringify(
```
Usually logs cost money and we tend to avoid logging every request. Instead, we generate arbitrary ids which helps us trace requests down the line. 

Another thing is that, right now you have a coupled middleware: Authentication + Logging. 
It's a good practice to have a separate middleware for handling the request logging or the creation of the above mentioned arbitrary id and appending it to the context.

ynaydenov-qb replied:
I decoupled them. In this application I didn't use express application and only apollo server, therefore those middlewares were combined in a function in app.ts files and passed as context. If you think that it is better idea to use express app and its middleware functionality I can do it.

mbozhilov-qb replied:
@astanilov correct me if I'm mistaken, but you are already running an express app the way you initialize your server. No need to explicitly define an express middleware. 

What you can do is stack middleware's. Both your logging middleware and your auth middleware should have access to the request, right?

```ts
createMiddlewareStack(req: Request) {
     loggingMiddleware(req);
     authMiddleware(req);
}
```

ynaydenov-qb replied:
I used more or less the same approach in my code:

```
const createContext = ({ req }: {req : Request}) => {
  // Initialize an empty context object
  let context = {};

  // Apply authentication and logging middleware
  context = { ...context, ...authenticationContextMiddleware({ req }) };
  context = { ...context, ...loggingContextMiddleware({ req }) };

  return context;
};
```

astanilov replied:
I might be late to the party, but yeah as far as I see, the apollo-server lib does use Express when creating the server

asynchroza commented on src/middleware/authentication.ts:
```diff
@@ -0,0 +1,22 @@
+  const expectedHeader: string = process.env.AUTH_TOKEN!;
+
+  if (authHeaderUser !== expectedHeader) {
+    logger.error(`Unauthorized access attempt: ${req.method} ${req.url}`);
```
I wouldn't really log this. Throwing the exception is good enough.

ynaydenov-qb replied:
Removed it, thanks!

astanilov commented on src/models/book.ts:
```diff
@@ -0,0 +1,26 @@
+import { LendingRecord } from "./lendingRecord";
+import { v4 as uuidv4 } from "uuid";
```
instead of importing this lib every place we need it, let's create an utils.ts file and create a method for creating IDs using that lib there. Then we can simply import the method and use it where needed

ynaydenov-qb replied:
Nice approach, I will implement it!

asynchroza commented on src/middleware/authentication.ts:
```diff
@@ -0,0 +1,22 @@
+
+export const authenticationContextMiddleware = ({ req }: { req: any }) => {
+  const authHeaderUser: string = req.headers.authorization || "";
+  const expectedHeader: string = process.env.AUTH_TOKEN!;
```
To avoid using ! here, it's better to load these environment variables in a separate file, such as environment.ts. In this file, you can create an env = {} object and a utility function that retrieves the environment variables. This function should throw an exception if a required value is missing. By handling potential errors at the beginning, you ensure that any issues with missing environment variables are caught early, preventing runtime problems later.

This is something a little bit more complex than what you [need](https://github.com/t3-oss/t3-env) but you can get inspired by it.

ynaydenov-qb replied:
I created an environment file where I check the environment variables. I made the use of the singleton design pattern as there should be only one environment and instanciated the variables in an object which I export.

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+
+    // Lend a book to a customer, updates the fields of a book related to lending
+    lendBook: (
+      _: any,
```
Help me remember, why do you have to drop the first argument here?

ynaydenov-qb replied:
If I remember correctly the first argument is the "root/parent" argument, in our case there is no relevant data passed down, therefore we don't need it for this mutation.

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+      const book = booksDataSource.getBookById(bookId);
+      const customer = customersDataSource.getCustomerById(customerId);
+
+      if (book && customer && !book.isLent) {
```
I highly suggest you read this [blog post](https://levelup.gitconnected.com/follow-these-6-patterns-or-i-will-reject-your-pull-request-fc08f908e7fe).

What we tend to do in order make our code more readable is use early returns. You can inverse your logic here and check for the bad paths first and throw the appropriate errors. This is called keeping the happy path on the left side.

ynaydenov-qb replied:
I read the article and it was very insightful. Now I refactored the code to follow this early return principle. This allowed for better error messages and is also easier to understand. Thanks!

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+
+        return book;
+      } else {
+        const errorMessage = `Book with ID ${bookId} is already lent or does not exist.`;
```
This is too much of an ambiguous message. I would suggest handling these separately.

ynaydenov-qb replied:
After following the "keep the happy path on the left side" idea all the error messages are more specific.

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+      _: any,
+      { bookId, returnDate }: { bookId: string; returnDate: string }
+    ): Book => {
+      const book = booksDataSource.getBookById(bookId);
```
My comment about the `early returns` could be applied to all of these.

ynaydenov-qb replied:
I applied it, thanks!

asynchroza commented on src/schemas/lendingRecordSchema.ts:
```diff
@@ -0,0 +1,11 @@
+    customer: Customer!
+    lentDate: String!
+    returnDate: String
+    dueDate: String!
```
Let's define Date as a new Scalar type. 

```
dueDate: Date!
```

ynaydenov-qb replied:
After a small fight with graphQL and JSON objects I did it

asynchroza commented on package.json:
```diff
@@ -0,0 +1,28 @@
+  "scripts": {
+    "start": "ts-node src/app.ts",
+    "build": "tsc",
+    "dev": "ts-node-dev src/app.ts"
```
`ts-node-dev` is not listed in dependencies. Also, `types` dependencies and `typescript` should be listed in dev dependencies because they're not dependencies used during runtime in production environments. 

https://issuu.com/johan.corrales.teams.test.1/docs/effective_typescript/s/16753105#:~:text=Because%20Type%E2%80%90Script%20is%20a,TypeScript%20generally%20belong%20in%20devDependencies.

ynaydenov-qb replied:
You are absolutely right, at the beginning I was confused what to add where, thanks for the explanation! I moved the dependencies to dev-dependecies.

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+  LendingRecord: {
+    // Find the customer associated with a lending record
+    customer: (record: LendingRecord): Customer => {
+      return customersDataSource.getCustomerById(record.customerId)!;
```
Hmmm. Something smells fishy here. Why do you have to use the assertion operator here?

ynaydenov-qb replied:
I used it because otherwise the return type would have been `Customer | undefined`, but if we have a lending record the only way to already have it is to have called the `lendBook` method, and it already has assigned valid customer to the lending record. So if I follow the logic if we have a lending record, we already have the customer in the database, however typescript can't be sure about it and therefore I ma using the assertion operator.

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+  Book: {
+    // Find the current customer that is lending the book
+    currentLendee: (book: Book): Customer | null => {
+      if (book.currentLendeeId)
```
Similarly to what we talked about a little above. Happy path to the left, bad path to right. 
```
if (!book.currentLendeeId) return null
...
return customersDataSource.getCustomerById(book.currentLendeeId);
```
is more readable imo.

ynaydenov-qb replied:
Absolutely right, I fixed it.

asynchroza commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+    // Find the current customer that is lending the book
+    currentLendee: (book: Book): Customer | null => {
+      if (book.currentLendeeId)
+        return customersDataSource.getCustomerById(book.currentLendeeId)!;
```
Also, I think you should think about the assertion operator here. I don't think it belongs here.
Typescript is trying to tell you something which is reasonable. You might have the id but this doesn't mean the user's account still exists.

ynaydenov-qb replied:
Note taken, I didn't think that it would be possible, but I added an if check.

mbozhilov-qb commented:
Somehow I've made all of my comments with my personal account. Excuse me 😞

mbozhilov-qb commented:
And I keep doing random stuff. The sickness has gotten to me and it's trying to sabotage not only my health but my career as well 👿 . Jokes aside. Amazing stuff @ynaydenov-qb! Pretty impressive for a first go at it. Keep up the good stuff.

mbozhilov-qb commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+    },
+  },
+
+  LendingRecord: {
```
Preferably, you should break these resolvers in a similar way to what Nasko suggested for models.

ynaydenov-qb replied:
I broke the resolvers into different files. Do you think it would be better it I instanciate the data sources in the app.ts file and then pass them as context if the user is authenticated, because in the current version I am instanciating an instance in every resolver?

mbozhilov-qb replied:
https://stackoverflow.com/a/69096721

ynaydenov-qb replied:
I moved the dataSources to the context.

mbozhilov-qb commented on src/dataSources/booksDataSource.ts:
```diff
@@ -0,0 +1,50 @@
+
+  // Add a new book
+  addBook(book: Book) {
+    this.books.push(book);
```
Hmm. What if it fails to save to file?

mbozhilov-qb replied:
This will stay in memory but upon server restart it will be discarded.

ynaydenov-qb replied:
You are write, now I am checking if writing to a file succeeds using a try/catch block and if not then I am reverting the change in memory as well.

asynchroza replied:
Is this the best approach in your case? You're accounting for the fact that this in memory object might be updated by the time you're removing the entry but you're doing an additional O(n) operation to find the entry. What are you gaining by pushing the object into the array before writing it to disk?

ynaydenov-qb replied:
You are right, I will just try to write it first and if it succeeds then I will modify the in memory object

mbozhilov-qb commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+
+  Book: {
+    // Find the current customer that is lending the book
+    currentLendee: (book: Book): Customer | null => {
```
I would love it if you figure out how to prove that these are not invoked when they're not queried for. Hint: make a slow read operation. 

This will help you wrap your head around why GQL is more efficient than regular Rest endpoints (in some cases).

As per the moment being, you won't really notice any difference in speed due to minimal size of the files, but once you have to start parsing larger json objects this will become pretty evident.

ynaydenov-qb replied:
I verified that these are not invoked when not asked for by adding a 2 seconds delay for obtaining this information and then manually testing in the apollo sandbox(a case when we query for them and case when not doing so). The code looked as follows:
![image](https://github.com/user-attachments/assets/d59862e2-507d-4b87-92c5-3ae765459481)

mbozhilov-qb replied:
Can't you just return the customer directly in this code? Is the null value something imposed by Apollo?
One of the big fights in the community is what null and undefined signify. Biased TLDR: they're the same with the weird difference being that if you do `typeof null` this is an object instead of a type signifying non-existent value. 

Hence, I usually always ask people to remove null from their code if possible especially when they have values which inherently are undefined and they have to cast them to nulls. 

I don't remember the Apollo spec here but I believe you can have this function return Promise<Customer | undefined> instead and rewrite the code as such:

```typescript
if (!book.currentLendeeId) return;
return await customerDataSource.getCustomerById(book.currentLendeeId); // that is if you have moved to asynchronous context
```

ynaydenov-qb replied:
I looked it up and I can have `Promise<Customer | undefined>`, therefore modified the methods to return that instead of null.

mbozhilov-qb commented on src/resolvers.ts:
```diff
@@ -0,0 +1,130 @@
+const booksDataSource = new BooksDataSource();
+const customersDataSource = new CustomersDataSource();
+
+export const resolvers = {
```
Most of your code is synchronous. Which is not an issue.. but that would rarely be the case in a real working environment. I would suggest you transform your code to use Promises so that you figure out how to resolve them.

Another thing I would suggest you think about is whether a deadlock may occur when writing to the same file.

ynaydenov-qb replied:
I made the code asynchronous. About the deadlock I don't think that a deadlock may occur since I am not using a lock and I am using now asynchronous read/write files operations. Furthermore If  I remember correctly we don't have many threads here and we can't have a situation where one thread locks a resource and depends on another one.

ynaydenov-qb committed:
Move detenv condiguration at top of app.ts and refactor logger

Loaded the environment variables at the start of the application. Changed the logs to be saved in json format and also started printing them in the console while in development mode.

ynaydenov-qb committed:
Removed import * as ... from packages

ynaydenov-qb committed:
Created a utils.ts where a function for generating uuids is exported

Created a utils.ts where a function for generating uuids is exported and used it in all classes where uuids were needed

ynaydenov-qb committed:
Forgot to add the utils.ts file and added it now

ynaydenov-qb committed:
Removed logging if an unauthorized access was denied

ynaydenov-qb committed:
Moved types and typescript dependencies to dev dependencies, locked versions

ynaydenov-qb committed:
Change the type of req in authentication from any to Request

ynaydenov-qb committed:
Refactored the code so it follows the "keep the happy path on the left side" idea

ynaydenov-qb committed:
Decouple authentication and logging from the authentication middleware

ynaydenov-qb committed:
Separate resolvers in different files and combine them finally

ynaydenov-qb committed:
Revert "Separate resolvers in different files and combine them finally"

This reverts commit da0e5daa1df9a9751ef8145cdbc106fe64ca4efd.

ynaydenov-qb committed:
Accidentally deleted app.ts and reverted commit

ynaydenov-qb committed:
Remove the assertion operator in obtaining the current lendee of a book

ynaydenov-qb committed:
Created an environment file where the environment variables are stored

ynaydenov-qb committed:
Added ESLint and Prettier in vsCode as well as created scripts for them

ynaydenov-qb committed:
Add a description in the README file

ynaydenov-qb committed:
Define Date as a scalar type

ynaydenov-qb committed:
Alter the code such that it uses promises

ynaydenov-qb committed:
Handle cases where the writing to a file fails

ynaydenov-qb committed:
Moved dataSources to the context

ynaydenov-qb committed:
Altered the methods to return undefined instead on null

ynaydenov-qb committed:
Save before editing in memory objects

astanilov commented on src/app.ts:
```diff
@@ -0,0 +1,35 @@
+  let context = {};
+
+  // Apply authentication and logging middleware
+  context = { ...context, ...authenticationContextMiddleware({ req }) };
+  context = { ...context, ...loggingContextMiddleware({ req }) };
```
We can safely create the middleware at once:
```suggestion
  context = { ...authenticationContextMiddleware({ req }), ...loggingContextMiddleware({ req }) };
```
And we don't even need to spread `context` here, is it will always be an empty object {} at this point.
We can even return the whole context with a one liner, but for the sake of readability let's keep it like that

ynaydenov-qb replied:
You are right, but now I added the dataSources to the context as well, therefore I think it is more readable on separate lines.

astanilov commented on src/customScalars/dateScalar.ts:
```diff
@@ -0,0 +1,28 @@
+export const DateScalar = new GraphQLScalarType({
+  name: 'Date',
+  description: 'Custom Date scalar type',
+  serialize(value: unknown): string {
```
`value: unknown` seems okay, but since we check if it is Date type, we can include that type in the parameter as `Date | unknown`

ynaydenov-qb replied:
Your suggestion makes sence, I included it in the code.

astanilov commented on src/dataSources/booksDataSource.ts:
```diff
@@ -0,0 +1,86 @@
+  async addBook(book: Book): Promise<void> {
+    try {
+      // First, save the new book to the file
+      await this.saveData();
+
+      // If saving is successful, update the in-memory data
+      this.books.push(book);
```
I am a bit confused here, since `saveData` is saving `this.books` to a file, shouldn't we push the new book before we save? Or we are purposely keeping in memory and file out of sync for some reason?

ynaydenov-qb replied:
You are right, I got messed up by the previous comment by Misho and have commited a mistake, I will fix it.

ynaydenov-qb replied:
I reverted to the previous version, where I first updated them in-memory objects and then saved them to file.

asynchroza replied:
I would suggest rethinking the flow here to be honest. Having `saveData` act on an array which could be accessed by multiple actors is not a good design decision. Saving a book should only save this one book. The in memory object shouldn't be used as the source of truth for mutations but rather as the cached entity for retrieving data for queries.

ynaydenov-qb replied:
I will think of a way to just append the new book to the file without reading it and then updating the in-memory array.

ynaydenov-qb replied:
After looking up in the internet what I found is that all responses suggest to read the data from the file, modify the data by adding a new object and then convert the data back to JSON string and save it. Won't this be slower than the current method?

asynchroza replied:
Yes, we will have an additional layer of json parsing but that's fine. You should never sacrifice data integrity.

ynaydenov-qb replied:
Then I will implement it.

ynaydenov-qb replied:
It is now done, please take a look and I am waiting for feedback. The methods about getting a single book/customer rely on the in-memory array, do you think that is a problem. I am updating that in memory array when adding or updating anything to the files.

mbozhilov-qb replied:
Yes, that's fine. We use the in-memory array as the cache. We should probably think of some type of invalidation but for the moment being that's totally fine. I think you've covered all the requirements and we can potentially move to other things.

astanilov commented on src/dataSources/booksDataSource.ts:
```diff
@@ -0,0 +1,86 @@
+    if (index !== -1) {
+      try {
+        await this.saveData();
+        this.books[index] = updatedBook;
```
ditto about in-memory and file sync

astanilov commented on src/dataSources/customersDataSource.ts:
```diff
@@ -0,0 +1,79 @@
+  async addCustomer(customer: Customer): Promise<void> {
+    try {
+      await this.saveData();
+      this.customers.push(customer);
```
ditto

astanilov commented on src/dataSources/customersDataSource.ts:
```diff
@@ -0,0 +1,79 @@
+    if (index !== -1) {
+      try {
+        await this.saveData();
+        this.customers[index] = updatedCustomer;
```
ditto

ynaydenov-qb committed:
Revert "Save before editing in memory objects"

This reverts commit eab34b92f73cdf34d810557bd0b5422d289e7935.

ynaydenov-qb committed:
Change type of the argument used in the serialize method for dateScalar

ynaydenov-qb committed:
Update the methods for adding and updating books and customers

Update the methods for adding and updating books and customers such that they read the file, parse it, add a new entry, stringify the new array and save it back to the file.
